code/job.py

```diff
@@ -0,0 +1,69 @@
+from flask_restful import Resource, reqparse
+
+jobs = []
+
+class Job(Resource):
+    parser = reqparse.RequestParser()
+    parser.add_argument(
+        'company',
+        type=str,
+        required=True,
+        help="This field cannot be left blank"
+        )
+    parser.add_argument('salary')
+    parser.add_argument('status')
+
+    def post(self, title):
+
+        if self.check_for_job(title) is not None:
+            return{'message': "A job with name '{}' already exists".format(title)}, 400
+
+        data = Job.parser.parse_args()
+
+        job = {
+            'title': title, 
+            'company': data['company'],
+            'salary': data['salary'],
+            'status': data['status']
+            }
+        jobs.append(job)
+
+        return job, 201
+
+    def check_for_job(self, title):
+        for job in jobs:
+            if job['title'] == title:
+                return job
+        return None
+
+    def get(self, title):
+        job = self.check_for_job(title)
+        if job:
+            return {'job': job}, 200
+        else:
+            return{"message": "Job does not exist"}, 404
+
+    def put(self, title):
+
+        data = Job.parser.parse_args()
+
+        job = self.check_for_job(title)
+        if job is None:
+            job = {
+                'title': title,
+                'company': data['company'],
+                'salary': data['salary'],
+                'status': data['status']
+                }
+            jobs.append(job), 201
```

Comment:
Ah, REST, how you are a monumental pain in the arse! I have no idea why the spec for REST states that you should use POST to create new items, but PUT is for updating them... unless you want to let it also create new items!

Personally, I would leave PUT for just updating existing items, as this then gives the PUT command itself more meaning. You are saying that this is for updates only, so then you can add checks to the method (return a 400 if the item doesn't exist). It also stops the accidental creation of new items because of a typo in the title.

You then end up with a smaller method that is doing the absolute minimum amount of work, so should theoretically be easier to understand.

Alternatively (though I wouldn't advise it), if you insist on the PUT method doing both things, then you might as well sack off the POST method, because what's the point of it?

Reply:
Not sure if it's a terrible idea, but did think the PUT could return POST method if the job doesn't already exist? i.e.: 
       ... if jobs.get(title):
            jobs[title].update(data), 200
            return jobs[title]
        else:
            return self.post(title)

Reply:
Is it a terrible idea? Not really. However, it feels a lot cleaner to have the method only do one thing.

My argument here would be, so why doesn't the POST method call the PUT method if the job already exists, rather than return an error? Sure, most definitions of REST state that the POST method should only create new items, not update existing ones, but why the difference with PUT then? I fear that REST descends in to ideological wankery far too easily; any protocol that it is deemed ok to end a discussion with "But that's not in accordance with the principles of REST" needs to have a long, hard look at itself!

Reply:
I agree it feels a lot cleaner to have each method only doing one thing, and I definitely didn't like the repetition needed for creation in both methods. I don;t think my alternative (calling POST within PUT) isn't great as it's a bit misleading. 

So simplifying PUT to just update or return an error (item not found) seems the cleanest option!

Reply:
It's not just cleaner, it's safer. I wouldn't want an update method where I can send the data up incorrectly (and mangle, or miss out, the id), and suddenly I've got a new entity in my data store. And I won't know that until I look in the data store.